src/arrays/spinops.jl

```diff
@@ -0,0 +1,110 @@
+#########################
+# Spin Operator Methods #
+#########################
+#########################
+# Implementation  -  1  #
+#########################
+#function spin_h1(j::Float64)
+#    if (isinteger(2*j) && j>0)
+#        m = reverse([-j:j-1])
+#        N = length(m)+1
+#        m_coeffs = [sqrt(j*(j+1.0) - (x+1.0)*x) for x in m]
+#        return spdiagm(m_coeffs,1,N,N)
+#    else
+#        return error("j must be non-negative integer or half integer")
+#  end
+#end
+
+#function spin_h2(j::Float64)
+#    if (isinteger(2*j) && j>0)
+#        m = reverse([-j:j])
+#        N = length(m)
+#        return spdiagm(m,0,N,N)
+#    else
+#        return error("j must be non-negative integer or half integer")
+#  end
+#end
+
+#spin_jx(j::Float64) = QuArray(0.5*(spin_h1(j)+spin_h1(j)'))
+#spin_jy(j::Float64) = QuArray(-0.5*im*(spin_h1(j)-spin_h1(j)'))
+#spin_jz(j::Float64) = QuArray(spin_h2(j))
+#spin_jp(j::Float64) = QuArray(spin_h1(j))
+#spin_jm(j::Float64) = QuArray(spin_h1(j)')
+#const sigma_x = 2.0 * spin_jx(0.5)
+#const sigma_y = 2.0 * spin_jy(0.5)
+#const sigma_z = 2.0 * spin_jz(0.5)
+
+# TODO : unicode sigma_y, sigma_z
+#const σₓ = sigma_x
+
+#########################
+# Implementation  -  2  #
+#########################
+#abstract Spin{S}
+
+immutable HalfSpin{S} end
+
+# constructors
+HalfSpin(j::Integer) = j > 0 ? HalfSpin{2*j}() : error("Spin must be positive")
+
+function HalfSpin(j::Float64)
+    if (j > 0) && isinteger(j)
+        HalfSpin(int(j))
+    elseif (j > 0) && isinteger(j*2)
+        HalfSpin{int(2*j)}()
+    else
+        error("Spin must be positive and a multiple of 1/2.")
+    end
+end
+
+# extracting the spin
+spin_value{S}(his::HalfSpin{S}) = S/2
+
+spin{S}(is::HalfSpin{S}) = is
+spin(j::Integer) = HalfSpin(j)
+spin(j::Float64) = HalfSpin(j)
+
+function mat_coeffs_1(j::Float64)
+    m = reverse([-j:j-1])
+    N = length(m)+1
+    m_coeffs = [sqrt(j*(j+1.0) - (x+1.0)*x) for x in m]
+    return spdiagm(m_coeffs,1,N,N)
+end
+
+function mat_coeffs_2(j::Float64)
+    m = reverse([-j:j])
+    N = length(m)
+    return spdiagm(m,0,N,N)
+end
+
```

Comment:
`reverse([-j:j-1])` above can be replaced by `[j-1:-1:-j]`, which should be more memory efficient since you'll only need to pass through the range once (`reverse` also works on ranges, btw: `reverse(1:10)` --> `10:-1:1`).